src/libs/actions/Card.ts

```diff
@@ -84,47 +91,53 @@ function reportVirtualExpensifyCardFraud(cardID: number) {
  * @param cardID - id of the card that is going to be replaced
  * @param reason - reason for replacement
  */
-function requestReplacementExpensifyCard(cardID: number, reason: ReplacementReason) {
-    const optimisticData: OnyxUpdate[] = [
-        {
-            onyxMethod: Onyx.METHOD.MERGE,
-            key: ONYXKEYS.FORMS.REPORT_PHYSICAL_CARD_FORM,
-            value: {
-                isLoading: true,
-                errors: null,
+function requestReplacementExpensifyCard(cardID: number, reason: ReplacementReason, validateCode: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+        const optimisticData: OnyxUpdate[] = [
+            {
+                onyxMethod: Onyx.METHOD.MERGE,
+                key: ONYXKEYS.ACCOUNT,
+                value: {
+                    isLoading: true,
+                },
             },
-        },
-    ];
+        ];
 
-    const successData: OnyxUpdate[] = [
-        {
-            onyxMethod: Onyx.METHOD.MERGE,
-            key: ONYXKEYS.FORMS.REPORT_PHYSICAL_CARD_FORM,
-            value: {
-                isLoading: false,
+        const successData: OnyxUpdate[] = [
+            {
+                onyxMethod: Onyx.METHOD.MERGE,
+                key: ONYXKEYS.ACCOUNT,
+                value: {
+                    isLoading: false,
+                },
             },
-        },
-    ];
+        ];
 
-    const failureData: OnyxUpdate[] = [
-        {
-            onyxMethod: Onyx.METHOD.MERGE,
-            key: ONYXKEYS.FORMS.REPORT_PHYSICAL_CARD_FORM,
-            value: {
-                isLoading: false,
+        const failureData: OnyxUpdate[] = [
+            {
+                onyxMethod: Onyx.METHOD.MERGE,
+                key: ONYXKEYS.ACCOUNT,
+                value: {
+                    isLoading: false,
+                },
             },
-        },
-    ];
+        ];
 
-    const parameters: RequestReplacementExpensifyCardParams = {
-        cardID,
-        reason,
-    };
+        const parameters: RequestReplacementExpensifyCardParams = {
+            cardID,
+            reason,
+            validateCode,
+        };
 
-    API.write(WRITE_COMMANDS.REQUEST_REPLACEMENT_EXPENSIFY_CARD, parameters, {
-        optimisticData,
-        successData,
-        failureData,
+        // eslint-disable-next-line rulesdir/no-api-side-effects-method
+        API.makeRequestWithSideEffects(SIDE_EFFECT_REQUEST_COMMANDS.REQUEST_REPLACEMENT_EXPENSIFY_CARD, parameters, {optimisticData, successData, failureData})
+            .then((response): string | undefined => {
+                if (response?.jsonCode !== CONST.JSON_CODE.SUCCESS) {
+                    return;
+                }
+                resolve((response as ExpensifyCardID).newCardID.toString());
+            })
+            .catch(() => reject());
     });
 }
```

Comment:
Returning promises from action methods is a discouraged pattern - it looks like we can do this without the side effects. 

We just want the new cardID here, so we could probably just have it update an onyx key and only re-direct if we have that key present and are done loading. Alternatively, it seems like we didn't have a redirect before so maybe we don't need to have it still

Reply:
@thienlnam without the side effects we used to redirect to the old card page and we display a "$0" limit which looks like a bug to the user


https://github.com/user-attachments/assets/0c440abf-cc94-4936-99bd-810e2a73c0ae

We also use side effects for reveal virtual card details here https://github.com/Expensify/App/blob/084b1047e8f090a63eb4eb99ab89d13f267bbe8b/src/libs/actions/Card.ts#L197. We also already display a loading spinner when requesting a new card.

Reply:
We use side effects there because we can't store those details in onyx since they're sensitive - but for this which is just a cardID it's not

Reply:
So how do you suggest fixing the UI glitch above? since we can't redirect to that page unless we have the new cardID. 

>Alternatively, it seems like we didn't have a redirect before so maybe we don't need to have it still

We did actually, but to the previous cardID that becomes `null` in Onyx so we display $0.

Reply:
👍 lmk if you need some ideas - usually the way we get around this is by adding a loading indicator on a loading key on the onyx key for the page

Reply:
@thienlnam I think I'm going to need an extra Web-E PR to set the newCardID key somewhere so I listen for it in the component then redirect to it once the backend responds. As otherwise there is no way of redirecting to the new card page and we'll end up showing the previous card with $0 limit

Reply:
Something like this, wdyt?

<img width="508" alt="Screenshot 2024-11-01 at 18 40 40" src="https://github.com/user-attachments/assets/e58c5f19-3a55-4460-85ca-a10160d80da8">

Reply:
@thienlnam IMO I think we can keep the side effect call here, you can see in the comment that our use case here is allowed, we are both calling Marqeta and redirecting based on the response

<img width="1177" alt="Screenshot 2024-11-01 at 19 10 59" src="https://github.com/user-attachments/assets/ea6cd922-6b65-4fdd-ab88-a5c251b95957">

Also given that we display a spinner and make the user wait, I don't see how is using the optimistic pattern is helpful?

Reply:
Yup something along those lines would work - you'd need to have the API command send an onyx update for the newCardID if you wanted to redirect to the new card.

Though thinking about this more though, if you only added the redirect so that the card doesn't show with a $0 limit  we could also just not do the redirect because then it forces this flow to happen online.

In App, in the optimistic data you could set a new key called isReplacing or something on the cardID and so when you open that cardID you would see some note that it's been replaced and then when the API completes it would remove the card and have the new card in the list. That's probably easier and then allows the user to take other actions offline with the command queued - what do you think?

Reply:
I hear you. However I think that this stuff is out of scope, so it's best if I kept this PR focused on validateCode using the same offline pattern we have now, which means not fixing the $0 bug, I think we can open an issue to fix separately. Sounds good?

Reply:
Yeah that works for me